conda_libmamba_solver/state.py

```diff
@@ -0,0 +1,1404 @@
+
+- ``specs``: a mapping of package names to its corresponding ``MatchSpec``
+  objects. These objects are passed to the actual Solver, hoping it will return
+  a solution.
+- ``records``: a mapping of package names to ``PackageRecord`` objects. It will
+  end up containing the list of package records that will compose the final state
+  of the prefix (the _solution_). Its default value is set to the currently installed
+  packages in the prefix. The solver will alter this list as needed to accommodate
+  the final solution.
+
+If the algorithm was not iterative, the sole purpose of the solver would be to turn
+the ``specs`` into ``records``. However, ``conda``'s logic will try to constrain the
+solution to mimic the initial state as much as possible to reduce the amount of
+changes in the prefix. Sometimes, the initial request is too constrained, which results
+in a number of conflicts. These conflicts are then stored in the ``conflicts`` mapping,
+which will determine which ``specs`` are relaxed in the next attempt. Additionally,
+``conda`` stores other solve artifacts:
+
+- ``for_history``: The explicitly requested specs in the command-line should end up
+  in the history. Some modifier flags can affect how this mapping is populated (e.g.
+  ``--update-deps``.)
+- ``neutered``: Pieces of history that were found to be conflicting in the future and
+  were annotated as such to avoid falling in the same conflict again.
+
+The mappings stored in ``SolverOutputState`` are backed by ``TrackedMap`` objects,
+which allow to keep the reasons _why_ those specs or records were added to the mappings,
+as well as richer logging for each action.
+"""
+
+# TODO: This module could be part of conda-core once if we refactor the classic logic
+
+from collections import defaultdict, MutableMapping
+from enum import Enum
+from itertools import chain
+from types import MappingProxyType
+from typing import Any, Hashable, Iterable, Type, Union, Optional, Tuple, Mapping
+from os import PathLike
+import logging
+import sys
+
+from conda import CondaError
+from conda._vendor.boltons.setutils import IndexedSet
+from conda.auxlib import NULL
+from conda.auxlib.ish import dals
+from conda.base.constants import DepsModifier, UpdateModifier
+from conda.base.context import context
+from conda.common.io import dashlist
+from conda.common.path import get_major_minor_version, paths_equal
+from conda.exceptions import PackagesNotFoundError, SpecsConfigurationConflictError
+from conda.history import History
+from conda.models.match_spec import MatchSpec
+from conda.models.records import PackageRecord
+from conda.models.prefix_graph import PrefixGraph
+from conda.core.index import _supplement_index_with_system
+from conda.core.prefix_data import PrefixData
+from conda.core.solve import get_pinned_specs
+
+logger = logging.getLogger(__name__)
+
+
+class TrackedMap(MutableMapping):
+    # TODO: I am sure there's a better place for this object; e.g. conda.models
```

Comment:
Should we add a `conda_libmamba_solver/models.py`?

Reply:
Good idea!